File: packages/image_crop_compress/image_crop_compress_ios/ios/image_crop_compress_ios/Sources/image_crop_compress_ios/ResizeEngine.swift
import Flutter
import UIKit
import CoreGraphics

public class ResizeEngine {
    public static func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let sourcePath = args["sourcePath"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "sourcePath is required", details: nil))
            return
        }
        
        let targetWidth = args["width"] as? Int
        let targetHeight = args["height"] as? Int
        let maintainAspectRatio = args["maintainAspectRatio"] as? Bool ?? true
        
        if targetWidth == nil && targetHeight == nil {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Must provide width or height", details: nil))
            return
        }
        
        DispatchQueue.global(qos: .userInitiated).async {
            guard let image = UIImage(contentsOfFile: sourcePath) else {
                DispatchQueue.main.async {
                    result(FlutterError(code: "DECODE_ERROR", message: "Failed to decode image at \(sourcePath)", details: nil))
                }
                return
            }
            
            let originalWidth = image.size.width
            let originalHeight = image.size.height
            
            var finalWidth = targetWidth != nil ? CGFloat(targetWidth!) : originalWidth
            var finalHeight = targetHeight != nil ? CGFloat(targetHeight!) : originalHeight
            
            if maintainAspectRatio {
                if targetWidth != nil && targetHeight == nil {
                    let ratio = finalWidth / originalWidth
                    finalHeight = originalHeight * ratio
                } else if targetHeight != nil && targetWidth == nil {
                    let ratio = finalHeight / originalHeight
                    finalWidth = originalWidth * ratio
                } else if targetWidth != nil && targetHeight != nil {
                    let ratioW = finalWidth / originalWidth
                    let ratioH = finalHeight / originalHeight
                    let minRatio = min(ratioW, ratioH)
                    finalWidth = originalWidth * minRatio
                    finalHeight = originalHeight * minRatio
                }
            }
            
            let newSize = CGSize(width: finalWidth, height: finalHeight)
            let format = UIGraphicsImageRendererFormat()
            format.scale = 1.0 // Output at exact pixel dimensions
            let renderer = UIGraphicsImageRenderer(size: newSize, format: format)
            
            let resizedImage = renderer.image { _ in
                image.draw(in: CGRect(origin: .zero, size: newSize))
            }
            
            let nsStringPath = sourcePath as NSString
            var extensionStr = nsStringPath.pathExtension.lowercased()
            if extensionStr.isEmpty { extensionStr = "jpg" }
            
            let tempDir = FileManager.default.temporaryDirectory.appendingPathComponent("image_crop_compress")
            do {
                try FileManager.default.createDirectory(at: tempDir, withIntermediateDirectories: true, attributes: nil)
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "IO_ERROR", message: "Failed to create temp directory", details: nil))
                }
                return
            }
            
            let fileName = "resize_\(UUID().uuidString).\(extensionStr)"
            let fileURL = tempDir.appendingPathComponent(fileName)
            
            var data: Data?
            var mimeType = "image/jpeg"
            
            if extensionStr == "png" {
                data = resizedImage.pngData()
                mimeType = "image/png"
            } else {
                data = resizedImage.jpegData(compressionQuality: 1.0)
            }
            
            guard let finalData = data else {
                DispatchQueue.main.async {
                    result(FlutterError(code: "ENCODE_ERROR", message: "Failed to encode resized image", details: nil))
                }
                return
            }
            
            do {
                try finalData.write(to: fileURL)
                let sizeInBytes = try fileURL.resourceValues(forKeys: [.fileSizeKey]).fileSize ?? finalData.count
                
                DispatchQueue.main.async {
                    result([
                        "path": fileURL.path,
                        "width": Int(newSize.width),
                        "height": Int(newSize.height),
                        "sizeInBytes": sizeInBytes,
                        "extension": extensionStr,
                        "mimeType": mimeType
                    ])
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "SAVE_ERROR", message: error.localizedDescription, details: nil))
                }
            }
        }
    }
}

File: image-crop-compress/image_crop_compress_android/android/src/main/kotlin/com/imageprocessor/image_crop_compress_android/ResizeEngine.kt
package com.imageprocessor.image_crop_compress_android

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileOutputStream
import java.util.UUID

object ResizeEngine {
    fun handle(call: MethodCall, result: MethodChannel.Result, context: Context) {
        val sourcePath = call.argument<String>("sourcePath")
        val targetWidth = call.argument<Int>("width")
        val targetHeight = call.argument<Int>("height")
        val maintainAspectRatio = call.argument<Boolean>("maintainAspectRatio") ?: true

        if (sourcePath == null) {
            result.error("INVALID_ARGUMENT", "sourcePath is required", null)
            return
        }

        if (targetWidth == null && targetHeight == null) {
            result.error("INVALID_ARGUMENT", "Must provide width or height", null)
            return
        }

        CoroutineScope(Dispatchers.IO).launch {
            try {
                val bitmap = BitmapFactory.decodeFile(sourcePath)
                    ?: throw Exception("Failed to decode image at $sourcePath")

                val originalWidth = bitmap.width
                val originalHeight = bitmap.height

                var finalWidth = targetWidth ?: originalWidth
                var finalHeight = targetHeight ?: originalHeight

                if (maintainAspectRatio) {
                    if (targetWidth != null && targetHeight == null) {
                        val ratio = targetWidth.toFloat() / originalWidth
                        finalHeight = (originalHeight * ratio).toInt()
                    } else if (targetHeight != null && targetWidth == null) {
                        val ratio = targetHeight.toFloat() / originalHeight
                        finalWidth = (originalWidth * ratio).toInt()
                    } else if (targetWidth != null && targetHeight != null) {
                        val ratioW = targetWidth.toFloat() / originalWidth
                        val ratioH = targetHeight.toFloat() / originalHeight
                        val minRatio = kotlin.math.min(ratioW, ratioH)
                        finalWidth = (originalWidth * minRatio).toInt()
                        finalHeight = (originalHeight * minRatio).toInt()
                    }
                }

                val resizedBitmap = Bitmap.createScaledBitmap(bitmap, finalWidth, finalHeight, true)

                val extension = File(sourcePath).extension.lowercase().let { 
                    if (it.isEmpty()) "jpg" else it 
                }

                val compressFormat = when (extension) {
                    "png" -> Bitmap.CompressFormat.PNG
                    "webp" -> if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.R)
                        Bitmap.CompressFormat.WEBP_LOSSY
                    else
                        @Suppress("DEPRECATION") Bitmap.CompressFormat.WEBP
                    else -> Bitmap.CompressFormat.JPEG
                }

                val tempDir = File(context.cacheDir, "image_crop_compress")
                if (!tempDir.exists()) tempDir.mkdirs()
                val outputFile = File(tempDir, "resize_${UUID.randomUUID()}.$extension")

                FileOutputStream(outputFile).use { out ->
                    resizedBitmap.compress(compressFormat, 100, out)
                }

                val sizeInBytes = outputFile.length()
                if (bitmap != resizedBitmap) {
                    bitmap.recycle()
                }
                resizedBitmap.recycle()

                val mimeType = when (compressFormat) {
                    Bitmap.CompressFormat.PNG -> "image/png"
                    Bitmap.CompressFormat.JPEG -> "image/jpeg"
                    else -> "image/webp"
                }

                withContext(Dispatchers.Main) {
                    result.success(mapOf(
                        "path" to outputFile.absolutePath,
                        "width" to finalWidth,
                        "height" to finalHeight,
                        "sizeInBytes" to sizeInBytes,
                        "extension" to extension,
                        "mimeType" to mimeType
                    ))
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    result.error("RESIZE_ERROR", e.message, null)
                }
            }
        }
    }
}
